### PR TITLE
add psp17receiver, update tests

### PR DESCRIPTION
**Pull request by VargSupercolony**

### CODE REVIEW

File: contracts/token/psp20/impls.rs [hunk 1]
```diff
@@ -228,9 +231,26 @@ pub trait PSP20: PSP20Storage {
...
+        match to_receiver.call_mut().on_psp17_received(Self::env().account_id(), from, value, data)
+            .fire()
+        {
+            Ok(result) => match result {
+                Ok(_) => (),
+                _ => panic!("{}", IPSP17ReceiverError::TransferRejected(String::from("The contract with `to` address does not accept tokens.")).as_ref()),
+            },
+            Err(e) => match e {
+                EnvError::NotCallable => (),
+                _ => panic!("{}", IPSP17ReceiverError::TransferRejected(String::from("Unknown error: call failed")).as_ref()),
```
Comment (xgreenx) [inline contracts/token/psp20/impls.rs:249]: You can use the `e` here=)

File: contracts/token/psp20/impls.rs [hunk 2]
```diff
@@ -228,9 +231,26 @@ pub trait PSP20: PSP20Storage {
...
     }
 
+    fn _do_safe_transfer_check(operator: AccountId, from: AccountId, to: AccountId, value: Balance, data: Vec<u8>) {
+        let mut to_receiver: PSP17Receiver = ink_env::call::FromAccountId::from_account_id(to);
+        match to_receiver.call_mut().on_psp17_received(Self::env().account_id(), from, value, data)
+            .fire()
+        {
+            Ok(result) => match result {
+                Ok(_) => (),
+                _ => panic!("{}", IPSP17ReceiverError::TransferRejected(String::from("The contract with `to` address does not accept tokens.")).as_ref()),
```
Comment (xgreenx) [inline contracts/token/psp20/impls.rs:245]: You also can use error from the call here

**Review by PierreOssun — COMMENTED** [review]
Can you also add the mock & integration tests ? (check my PR)
The rest is Ok for me

**Comment by VargSupercolony** [comment]
yes ok, good idea, I will do that after your tests are merged to main

**Review by xgreenx — APPROVED** [review]

**Pull request merged**